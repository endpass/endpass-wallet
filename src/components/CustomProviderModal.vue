<template>
  <div class="new-account-modal">
    <modal-component @close="close">
      <template slot="header">Add New Provider</template>

      <div v-if="!providerAdded">
	      <form>
          <div class="field">
            <label class="label" for="name">Network name</label>
            <div class="control">
              <input v-model="provider.name" name="name" v-validate="'required'" type="text" class="input" :class="{'is-danger': fields.name && fields.name.touched && fields.name.invalid }" id="name" aria-describedby="name" placeholder="Network name" required>
            </div>
            <p class="help is-danger">{{errors.first('name')}}</p>
          </div>
          <div class="field">
            <label class="label" for="name">Provider url</label>
            <div class="control">
              <input v-model="provider.url" name="url" v-validate="'required|url:require_protocol:true|not_in:' + providersLinks" type="text" class="input" :class="{'is-danger': fields.url && fields.url.touched && fields.url.invalid }" id="name" aria-describedby="url" placeholder="Provider url" required>
            </div>
            <p class="help is-danger">{{errors.first('url')}}</p>
          </div>
	      </form>
      </div>
      <div v-else>
        <p class="subtitle">New Provider Added</p>

        <div class="message">
          <div class="message-header">
            <p>Provider Address</p>
          </div>
          <div class="message-body">
            <p>{{provider.name}}</p>
            <p class="code address">{{provider.url}}</p>
          </div>
        </div>
      </div>

      <div slot="footer">
        <div v-if="!providerAdded">
          <a class="button is-primary" @click="addNewProvider">Create
            New Provider</a>
          <a class="button" @click="close">Cancel</a>
        </div>
        <div v-else>
          <a class="button is-primary" @click="close">Close</a>
        </div>
      </div>

    </modal-component>
  </div>
</template>
<script>
import ModalComponent from '@/components/ModalComponent'
export default {
	data() {
		return {
			providerAdded:false,
			provider: {
				name: '',
				url: ''
			}
		}
	},
  computed: {
    providersLinks() {
      let links = this.$store.getters['web3/networks'].map(net => net.url).toString();
      return links
    }
  },
	methods: {
		addNewProvider() {
      this.$store.dispatch('web3/addNewProvider', this.provider);
      this.providerAdded = true;
		},
    close() {
      this.$emit('close')
    }
	},
  components: {
    ModalComponent
  }
}
</script>
<style lang="css">
	
</style>