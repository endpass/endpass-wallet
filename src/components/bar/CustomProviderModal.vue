<template>
  <div class="new-account-modal">
    <v-modal @close="close">
      <header slot="header">{{ headerText }}</header>
      <div v-if="!providerAdded">
        <v-form
          :is-form-valid="isFormValid"
          @submit="handleButtonClick"
        >

          <v-input
            v-validate="'required'"
            id="name"
            v-model="innerProvider.name"
            :disabled="isLoading"
            :error="errors.first('name')"
            name="name"
            data-vv-name="name"
            label="Network name"
            aria-describedby="name"
            placeholder="Network name"
            data-vv-as="Network name"
            autofocus
            required
          />

          <v-input
            v-validate="`required|url:require_protocol:true|not_in:${providersLinks}`"
            id="url"
            v-model="innerProvider.url"
            :disabled="isLoading"
            :error="errors.first('url')"
            data-vv-name="url"
            name="url"
            label="Provider url"
            aria-describedby="url"
            placeholder="Provider url"
            data-vv-as="Provider url"
            @input="handleInput"
          />

          <v-select
            v-validate="'required'"
            id="currency"
            :error="errors.first('currency')"
            v-model="innerProvider.currency"
            :options="currencies"
            data-vv-name="currency"
            name="currency"
            label="Provider currency"
            aria-describedby="currency"
            placeholder="Provider currency"
            data-vv-as="Provider currency"
            required
          />
        </v-form>
      </div>
      <div v-else>
        <p class="subtitle">{{ headerTextAfterAction }}</p>

        <div class="message">
          <div class="message-header">
            <p>Provider Address</p>
          </div>
          <div class="message-body">
            <p>{{ innerProvider.name }}</p>
            <p class="code address">{{ innerProvider.url }}</p>
          </div>
        </div>
      </div>
      <template slot="footer">
        <div class="buttons">
          <v-button
            v-if="!providerAdded"
            :loading="isLoading"
            :disabled="!isFormValid"
            class-name="is-primary is-medium"
            type="button"
            data-test="add-provider-button"
            @click="handleButtonClick"
          >
            {{ buttonText }}
          </v-button>
        </div>
      </template>

    </v-modal>
  </div>
</template>
<script>
import { mapActions, mapGetters } from 'vuex';
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VSelect from '@/components/ui/form/VSelect';
import VButton from '@/components/ui/form/VButton';
import { CURRENCIES } from '@/constants';
import formMixin from '@/mixins/form';

const defaultProvider = {
  name: '',
  url: '',
  currency: 1,
};

export default {
  name: 'CustomProviderModal',
  props: {
    provider: {
      type: Object,
      default: () => defaultProvider,
    },
  },
  data() {
    return {
      providerAdded: false,
      isLoading: false,
      innerProvider: Object.assign({}, this.provider),
      currencies: CURRENCIES.map(currency => ({
        val: currency.id,
        text: currency.name,
      })),
    };
  },
  computed: {
    ...mapGetters('web3', ['networks']),
    providersLinks() {
      const networks = this.needUpdateProvider
        ? this.networks.filter(net => net.url !== this.provider.url)
        : this.networks;

      return networks.map(net => net.url).toString();
    },
    needUpdateProvider() {
      return !Object.is(this.$props.provider, defaultProvider);
    },
    headerText() {
      return this.needUpdateProvider ? 'Update Provider' : 'Add New Provider';
    },
    buttonText() {
      return this.needUpdateProvider
        ? 'Update Provider'
        : 'Create New Provider';
    },
    headerTextAfterAction() {
      return this.needUpdateProvider
        ? 'Provider Updated'
        : 'New Provider Added';
    },
  },
  methods: {
    ...mapActions('web3', ['addNetwork', 'validateNetwork', 'updateNetwork']),
    async handleButtonClick() {
      // When add - providersLinks includes network => validation error
      this.$validator.pause();
      this.isLoading = true;

      try {
        const [networkType, networkId] = await this.validateNetwork({
          network: this.innerProvider,
        });
        const network = {
          ...this.innerProvider,
          id: networkId,
          networkType,
        };
        let isSuccess;

        if (this.needUpdateProvider) {
          isSuccess = await this.updateNetwork({
            network,
            oldNetwork: this.provider,
          });
        } else {
          isSuccess = await this.addNetwork({ network });
        }

        if (isSuccess) {
          this.providerAdded = true;
        }
      } catch (error) {
        this.errors.add({
          field: 'url',
          msg: 'Provider is invalid',
          id: 'wrongUrl',
        });
      } finally {
        this.$validator.resume();
        this.isLoading = false;
      }
    },
    close() {
      if (!this.isLoading) {
        this.$emit('close');
      }
    },
    handleInput() {
      this.errors.removeById('wrongUrl');
    },
  },
  mixins: [formMixin],
  components: {
    VModal,
    VInput,
    VSelect,
    VForm,
    VButton,
  },
};
</script>
<style lang="css">
</style>
