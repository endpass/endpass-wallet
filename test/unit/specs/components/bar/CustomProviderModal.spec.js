import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import { shallow, createLocalVue } from '@vue/test-utils';

import CustomProviderModal from '@/components/bar/CustomProviderModal';

describe('CustomProviderModal', () => {
  const web3Actions = {
    addNetwork: jest.fn(),
    validateNetwork: jest.fn(),
    updateNetwork: jest.fn(),
  };
  const web3Getters = {
    networks: jest.fn(() => [{ url: 'provider url' }]),
  };
  let wrapper;
  let componentOptions;

  beforeEach(() => {
    const localVue = createLocalVue();

    localVue.use(Vuex);
    localVue.use(VeeValidate);

    const store = new Vuex.Store({
      modules: {
        web3: {
          namespaced: true,
          actions: web3Actions,
          getters: web3Getters,
          state: {
            currencies: [{ id: 1, name: 'ETH' }],
          },
        },
      },
    });

    componentOptions = {
      store,
      localVue,
    };
  });

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallow(CustomProviderModal, {
        ...componentOptions,
      });
    });

    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('CustomProviderModal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    describe('url field', () => {
      it('should render correctly', () => {
        expect(wrapper.find('#url').html()).toMatchSnapshot();
      });
    });

    describe('header', () => {
      it('should render header for create new provider', () => {
        wrapper.setProps({});

        expect(wrapper.find('header').text()).toBe('Add New Provider');
      });

      it('should render header for update provider', () => {
        wrapper.setProps({
          provider: {},
        });

        expect(wrapper.find('header').text()).toBe('Update Provider');
      });
    });

    describe('footer', () => {
      it('should render button text for create new provider', () => {
        wrapper.setProps({});

        expect(wrapper.find('[data-test=add-provider-button]').text()).toBe(
          'Create New Provider',
        );
      });

      it('should render button text for update provider', () => {
        wrapper.setProps({
          provider: {},
        });

        expect(wrapper.find('[data-test=add-provider-button]').text()).toBe(
          'Update Provider',
        );
      });
    });
  });

  describe('computed', () => {
    beforeEach(() => {
      wrapper = shallow(CustomProviderModal, componentOptions);
    });

    describe('providersLinks', () => {
      const providerUrl1 = 'provider url 1';
      const providerUrl2 = 'provider url 2';

      it('should return provider links when creating a new provider', () => {
        let expected;

        wrapper.setProps({});
        wrapper.setComputed({
          networks: [{ url: providerUrl1 }],
        });
        expected = providerUrl1;

        expect(wrapper.vm.providersLinks).toBe(expected);

        wrapper.setComputed({
          networks: [{ url: providerUrl1 }, { url: providerUrl2 }],
        });
        expected = `${providerUrl1},${providerUrl2}`;

        expect(wrapper.vm.providersLinks).toBe(expected);
      });

      it('should return provider links when provider updates', () => {
        let expected;

        wrapper.setProps({
          provider: {
            url: providerUrl1,
          },
        });
        wrapper.setComputed({
          networks: [{ url: providerUrl1 }],
        });
        expected = '';

        expect(wrapper.vm.providersLinks).toBe(expected);

        wrapper.setComputed({
          networks: [{ url: providerUrl1 }, { url: providerUrl2 }],
        });
        expected = providerUrl2;

        expect(wrapper.vm.providersLinks).toBe(expected);
      });
    });
  });

  describe('methods', () => {
    beforeEach(() => {
      wrapper = shallow(CustomProviderModal, componentOptions);
    });

    describe('handleButtonClick', () => {
      const provider = {
        name: 'provider url',
        url: 'provider url',
        currency: 1,
      };
      const error = {
        field: 'url',
        msg: 'Provider is invalid',
        id: 'wrongUrl',
      };

      describe('creating a new provider', () => {
        beforeEach(() => {
          wrapper.setProps({});
          wrapper.setData({
            innerProvider: provider,
          });
        });

        it('should create new provider', async () => {
          const networkType = 'ropsten';
          const networkId = 3;
          const isSuccess = true;

          wrapper.setMethods({
            addNetwork: jest.fn().mockResolvedValueOnce(isSuccess),
            validateNetwork: jest
              .fn()
              .mockResolvedValueOnce([networkType, networkId]),
          });

          expect.assertions(6);

          await wrapper.vm.handleButtonClick();

          expect(wrapper.vm.validateNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.validateNetwork).toHaveBeenCalledWith({
            network: provider,
          });

          expect(wrapper.vm.addNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.addNetwork).toHaveBeenCalledWith({
            network: {
              ...provider,
              networkType,
              id: networkId,
            },
          });

          expect(wrapper.vm.isLoading).toBeFalsy();
          expect(wrapper.vm.providerAdded).toBe(isSuccess);
        });

        it('should handle the validation error of the network', async () => {
          wrapper.setMethods({
            validateNetwork: jest.fn().mockRejectedValue(),
          });
          spyOn(wrapper.vm.errors, 'add');

          expect.assertions(5);

          await wrapper.vm.handleButtonClick();

          expect(wrapper.vm.validateNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.validateNetwork).toHaveBeenCalledWith({
            network: provider,
          });
          expect(wrapper.vm.isLoading).toBeFalsy();
          expect(wrapper.vm.errors.add).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.errors.add).toHaveBeenCalledWith(error);
        });
      });

      describe('provider update', () => {
        beforeEach(() => {
          wrapper.setProps({
            provider,
          });
          wrapper.setData({
            innerProvider: provider,
          });
        });

        it('should update provider', async () => {
          const networkType = 'ropsten';
          const networkId = 3;
          const isSuccess = true;

          wrapper.setMethods({
            updateNetwork: jest.fn().mockResolvedValueOnce(isSuccess),
            validateNetwork: jest
              .fn()
              .mockResolvedValue([networkType, networkId]),
          });

          expect.assertions(6);

          await wrapper.vm.handleButtonClick();

          expect(wrapper.vm.validateNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.validateNetwork).toHaveBeenCalledWith({
            network: provider,
          });

          expect(wrapper.vm.updateNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.updateNetwork).toHaveBeenCalledWith({
            network: {
              ...provider,
              id: networkId,
              networkType,
            },
            oldNetwork: provider,
          });

          expect(wrapper.vm.isLoading).toBeFalsy();
          expect(wrapper.vm.providerAdded).toBeTruthy();
        });

        it('should handle the validation error of the network', async () => {
          wrapper.setMethods({
            validateNetwork: jest.fn().mockRejectedValue(),
          });
          spyOn(wrapper.vm.errors, 'add');

          expect.assertions(5);

          await wrapper.vm.handleButtonClick();

          expect(wrapper.vm.validateNetwork).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.validateNetwork).toHaveBeenCalledWith({
            network: provider,
          });
          expect(wrapper.vm.isLoading).toBeFalsy();
          expect(wrapper.vm.errors.add).toHaveBeenCalledTimes(1);
          expect(wrapper.vm.errors.add).toHaveBeenCalledWith(error);
        });
      });
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      wrapper = shallow(CustomProviderModal, componentOptions);
    });

    it('should emit close event when not loading', () => {
      wrapper.vm.close();

      wrapper.setData({
        isLoading: true,
      });

      wrapper.vm.close();

      expect(wrapper.emitted().close.length).toBe(1);
    });

    it('should remove an error when entering a new value', () => {
      wrapper.vm.errors.add({
        field: 'url',
        msg: 'Provider is invalid',
        id: 'wrongUrl',
      });

      expect(wrapper.vm.errors.has('url')).toBeTruthy();

      wrapper.vm.handleInput();

      expect(wrapper.vm.errors.has('url')).toBeFalsy();
    });
  });
});
