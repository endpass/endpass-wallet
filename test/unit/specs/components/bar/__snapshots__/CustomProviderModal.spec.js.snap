// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomProviderModal render url field should render correctly 1`] = `
<v-input id="url" data-vv-name="url" name="url" label="Provider url" aria-describedby="url" placeholder="Provider url" data-vv-as="Provider url" v-validate="\`required|url:require_protocol:true|not_in:\${providersLinks}\`" v-on:input="bound handleInput"
    v-model="innerProvider.url" aria-required="true" aria-invalid="false"></v-input>
`;
